param.go

```diff
@@ -98,9 +98,36 @@ type paramVisitor interface {
 // recursed into.
 type paramVisitorFunc func(param) (recurse bool)
 
-func (f paramVisitorFunc) Visit(p param) paramVisitor {
-	if f(p) {
-		return f
+// paramVisitOnce encapsulates a paramVisitorFunc and a map to track visited nodes.
+type paramVisitOnce struct {
+	f       paramVisitorFunc
+	visited map[string]struct{}
+}
+
+// NewParamVisitOnce constructs a paramVisitor avoiding redundant visits.
+func newParamVisitOnce(visited map[string]struct{}, f paramVisitorFunc) paramVisitor {
+	return paramVisitOnce{
+		f:       f,
+		visited: visited,
+	}
+}
+
+// Visit operates pv.f on param if:
+//		- param is not an edge node (i.e. param is a paramList), OR
+//		- param is an edge node that has not yet been visited
+func (pv paramVisitOnce) Visit(p param) paramVisitor {
+	switch p.(type) {
+	case paramSingle, paramGroupedSlice:
+		if _, ok := pv.visited[p.String()]; ok {
+			return pv
+		}
+		defer func() {
+			pv.visited[p.String()] = struct{}{}
```

Comment:
Why the defer? Wouldn't marking that param as visited right here before you
call `f` work?

Reply:
Hm maybe you are right, not sure why I thought it was important to only do it after `f`. Might be a vestige from before I was only returning early for edge parameters.